subsys/bluetooth/mesh/proxy_cli.c

```diff
@@ -0,0 +1,314 @@
+/*
+ * Copyright (c) 2021 Xiaomi Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/byteorder.h>
+
+#include <net/buf.h>
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/hci.h>
+#include <bluetooth/uuid.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/mesh.h>
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_MESH_DEBUG_PROXY)
+#define LOG_MODULE_NAME bt_mesh_proxy_client
+#include "common/log.h"
+
+#include "mesh.h"
+#include "adv.h"
+#include "net.h"
+#include "rpl.h"
+#include "transport.h"
+#include "host/ecc.h"
+#include "prov.h"
+#include "beacon.h"
+#include "foundation.h"
+#include "access.h"
+#include "proxy.h"
+#include "gatt_cli.h"
+#include "proxy_msg.h"
+
+static struct bt_mesh_proxy_server {
+	struct bt_mesh_proxy_role *srv;
+	bool link_opened;
+} servers[CONFIG_BT_MAX_CONN];
+
+static uint16_t allow_net_idx[CONFIG_BT_MAX_CONN] = {
+	[0 ... (CONFIG_BT_MAX_CONN - 1)] = BT_MESH_KEY_UNUSED,
+};
+
+static uint16_t *find_net_idx(uint16_t net_idx)
+{
+	for (int i = 0; i < ARRAY_SIZE(allow_net_idx); i++) {
+		if (allow_net_idx[i] == net_idx) {
+			return &allow_net_idx[i];
+		}
+	}
+
+	return NULL;
+}
+
+static void buf_send_end(struct bt_conn *conn, void *user_data)
+{
+	struct net_buf *buf = user_data;
+
+	net_buf_unref(buf);
+}
+
+bool bt_mesh_proxy_client_relay(struct net_buf *buf)
+{
+	bool relayed = false;
+	int i, err;
+
+	for (i = 0; i < ARRAY_SIZE(servers); i++) {
+		struct bt_mesh_proxy_server *server = &servers[i];
+
+		NET_BUF_SIMPLE_DEFINE(msg, 32);
+
+		if (!server->srv || !server->link_opened) {
+			continue;
+		}
+
+		/* Proxy PDU sending modifies the original buffer,
+		 * so we need to make a copy.
+		 */
+		net_buf_simple_reserve(&msg, 1);
+		net_buf_simple_add_mem(&msg, buf->data, buf->len);
+
+		err = bt_mesh_proxy_msg_send(server->srv->conn, BT_MESH_PROXY_NET_PDU,
+					     &msg, buf_send_end, net_buf_ref(buf));
+
+		bt_mesh_adv_send_start(0, err, BT_MESH_ADV(buf));
+		if (err) {
+			BT_ERR("Failed to send proxy message (err %d)", err);
+
+			/* If segment_and_send() fails the buf_send_end() callback will
+			 * not be called, so we need to clear the user data (net_buf,
+			 * which is just opaque data to segment_and send) reference given
+			 * to segment_and_send() here.
+			 */
+			net_buf_unref(buf);
+			continue;
+		}
+
+		relayed = true;
+	}
+
+	return relayed;
+}
+
+static int subnet_find_cb(struct bt_mesh_subnet *sub, void *cb_data)
+{
+	struct net_buf_simple *buf = (struct net_buf_simple *)cb_data;
+
+	if (!memcmp(sub->keys[0].net_id, buf->data, 8)) {
+		return 1;
+	}
+
+	if (!bt_mesh_subnet_has_new_key(sub)) {
+		return 0;
+	}
+
+	if (!memcmp(sub->keys[1].net_id, buf->data, 8)) {
+		return 1;
+	}
+
+	return 0;
+}
```

Comment:
Perhaps you could move this lookup into subnet.c, as beacon.c is also doing it.
E.g.
```c

static int has_net_id(struct bt_mesh_subnet * sub, void *net_id)
{
	return (!memcmp(sub->keys[0].net_id, net_id, 8)) ||
	       (bt_mesh_subnet_has_new_key(sub) && !memcmp(sub->keys[1].net_id, net_id, 8));
}

struct bt_mesh_subnet *bt_mesh_subnet_from_net_id(const uint8_t net_id[8])
{
	return bt_mesh_subnet_find(has_net_id, net_id);
}
```